File: web-simulator/js/shell.js
/**
 * TinyOLED Desktop — Desktop Shell (57-app browser catalog)
 * State machine: SPLASH -> HOME -> APP
 */
import { Framebuffer } from './framebuffer.js';
import { OLEDDisplay } from './display.js';
import { InputManager, EVT_UP, EVT_DOWN, EVT_SEL, EVT_LONG } from './input.js';
import { Scheduler } from './scheduler.js';
import { StatusBar } from './statusbar.js';
import { Launcher } from './launcher.js';
import { NotificationManager } from './notification.js';

import { ClockApp } from './apps/clock.js';
import { SysInfoApp } from './apps/sysinfo.js';
import { SettingsApp } from './apps/settings.js';
import { SnakeApp } from './apps/snake.js';
import { FlappyApp } from './apps/flappy.js';
import { DiceApp } from './apps/dice.js';
import { StarfieldApp } from './apps/starfield.js';
import { GameOfLifeApp } from './apps/gameoflife.js';
import { DVDLogoApp } from './apps/dvd.js';
import { MatrixRainApp } from './apps/matrix.js';
import { Cube3DApp } from './apps/cube3d.js';
import { FractalApp } from './apps/fractal.js';
import { MoonApp } from './apps/moon.js';
import { PomodoroApp } from './apps/pomodoro.js';
import { BreathingApp } from './apps/breathing.js';
import {
  WiFiApp, FileBrowserApp, PowerApp, TamagotchiApp, GraphApp, DockerApp, SystemdApp,
  GPIOViewerApp, I2CScannerApp, CommandRunnerApp, OscilloscopeApp, MultimeterApp,
  SSHAlertApp, TOTPApp, PasswordGeneratorApp, WiFiScannerApp, CryptoTickerApp,
  QRCodeApp, GitHubTrackerApp, SpeedtestApp, HackerNewsApp, TemperatureHumidityApp,
  UPSBatteryApp, PlantMonitorApp, CompassApp, MP3PlayerApp, RadioApp, VideoPlayerApp,
  TelegramApp, EmailApp, WorldClockApp, PixelArtApp, ScreenshotApp, HIITTimerApp,
  ServoControlApp, RobotCarApp, PiHoleApp, IPCameraApp, VoiceControlApp, SDHealthApp,
  APTUpdateApp, TodoApp
} from './apps/extended_apps.js';

const State = { SPLASH: 0, HOME: 1, APP: 2 };
const SPLASH_FRAMES = 40;
const TARGET_FPS = 20;

export class Desktop {
  constructor(canvas) {
    this._display = new OLEDDisplay(canvas);
    this._fb = new Framebuffer();
    this._state = State.SPLASH;
    this._app = null;
    this._splashT = 0;
    this._statusbar = new StatusBar();
    this._launcher = new Launcher();
    this._notif = new NotificationManager();
    this._sched = new Scheduler();
    this._input = new InputManager();
    this._onStateChange = null;
    this._registerApps(); this._registerTasks(); this._bindButtons();
  }
  set onStateChange(fn) { this._onStateChange = fn; }
  _registerApps() {
    const ex = () => this._goHome();
    const nx = (msg, dur) => this._notif.push(msg, dur);
    this._launcher.register('clock','Saat','clock',()=>this._open(new ClockApp(ex)));
    this._launcher.register('sysinfo','Sistem','cpu',()=>this._open(new SysInfoApp(ex)));
    this._launcher.register('wifi','WiFi','radar',()=>this._open(new WiFiApp(ex)));
    this._launcher.register('file','Dosya','script',()=>this._open(new FileBrowserApp(ex)));
    this._launcher.register('settings','Ayar','gear',()=>this._open(new SettingsApp(ex,nx)));
    this._launcher.register('power','Guc','service',()=>this._open(new PowerApp(ex)));
    this._launcher.register('snake','Yilan','snake',()=>this._open(new SnakeApp(ex)));
    this._launcher.register('flappy','Flappy','bird',()=>this._open(new FlappyApp(ex)));
    this._launcher.register('tamagotchi','Pet','pet',()=>this._open(new TamagotchiApp(ex)));
    this._launcher.register('dice','Zar','dice',()=>this._open(new DiceApp(ex)));
    this._launcher.register('graph','Grafik','graph',()=>this._open(new GraphApp(ex)));
    this._launcher.register('docker','Docker','docker',()=>this._open(new DockerApp(ex)));
    this._launcher.register('systemd','Svc','service',()=>this._open(new SystemdApp(ex)));
    this._launcher.register('gpio','GPIO','pin',()=>this._open(new GPIOViewerApp(ex)));
    this._launcher.register('i2c','I2C','i2c',()=>this._open(new I2CScannerApp(ex)));
    this._launcher.register('script','Komut','script',()=>this._open(new CommandRunnerApp(ex)));
    this._launcher.register('scope','Scope','scope',()=>this._open(new OscilloscopeApp(ex)));
    this._launcher.register('meter','Metre','volt',()=>this._open(new MultimeterApp(ex)));
    this._launcher.register('ssh','SSH','shield',()=>this._open(new SSHAlertApp(ex)));
    this._launcher.register('totp','TOTP','key',()=>this._open(new TOTPApp(ex)));
    this._launcher.register('passgen','Sifre','passkey',()=>this._open(new PasswordGeneratorApp(ex)));
    this._launcher.register('wifiscan','WiScan','radar',()=>this._open(new WiFiScannerApp(ex)));
    this._launcher.register('starfield','Yldiz','star',()=>this._open(new StarfieldApp(ex)));
    this._launcher.register('life','Life','cell',()=>this._open(new GameOfLifeApp(ex)));
    this._launcher.register('dvd','DVD','dvd',()=>this._open(new DVDLogoApp(ex)));
    this._launcher.register('matrix','Matrx','matrix',()=>this._open(new MatrixRainApp(ex)));
    this._launcher.register('crypto','Kripto','crypto',()=>this._open(new CryptoTickerApp(ex)));
    this._launcher.register('cube3d','3D','cube',()=>this._open(new Cube3DApp(ex)));
    this._launcher.register('fractal','Fraktl','fractal',()=>this._open(new FractalApp(ex)));
    this._launcher.register('qr','QR','qr',()=>this._open(new QRCodeApp(ex)));
    this._launcher.register('github','GitHub','github',()=>this._open(new GitHubTrackerApp(ex)));
    this._launcher.register('speed','Speed','speed',()=>this._open(new SpeedtestApp(ex)));
    this._launcher.register('news','HNews','news',()=>this._open(new HackerNewsApp(ex)));
    this._launcher.register('temp','Temp','temp',()=>this._open(new TemperatureHumidityApp(ex)));
    this._launcher.register('ups','UPS','ups',()=>this._open(new UPSBatteryApp(ex)));
    this._launcher.register('plant','Bitki','plant',()=>this._open(new PlantMonitorApp(ex)));
    this._launcher.register('compass','Pusula','compass',()=>this._open(new CompassApp(ex)));
    this._launcher.register('mp3','MP3','music',()=>this._open(new MP3PlayerApp(ex)));
    this._launcher.register('radio','Radyo','radio',()=>this._open(new RadioApp(ex)));
    this._launcher.register('video','Video','video',()=>this._open(new VideoPlayerApp(ex)));
    this._launcher.register('telegram','Tele','telegram',()=>this._open(new TelegramApp(ex)));
    this._launcher.register('email','Email','email',()=>this._open(new EmailApp(ex)));
    this._launcher.register('moon','Ay','moon',()=>this._open(new MoonApp(ex)));
    this._launcher.register('world','Dunya','world',()=>this._open(new WorldClockApp(ex)));
    this._launcher.register('pixel','Pixel','paint',()=>this._open(new PixelArtApp(ex)));
    this._launcher.register('shot','Shot','camera',()=>this._open(new ScreenshotApp(ex)));
    this._launcher.register('pomodoro','Pomo','timer',()=>this._open(new PomodoroApp(ex,nx)));
    this._launcher.register('breath','Nefes','breath',()=>this._open(new BreathingApp(ex)));
    this._launcher.register('hiit','HIIT','workout',()=>this._open(new HIITTimerApp(ex)));
    this._launcher.register('servo','Servo','servo',()=>this._open(new ServoControlApp(ex)));
    this._launcher.register('car','Araba','car',()=>this._open(new RobotCarApp(ex)));
    this._launcher.register('pihole','PiHole','pihole',()=>this._open(new PiHoleApp(ex)));
    this._launcher.register('ipcam','IPCam','camera',()=>this._open(new IPCameraApp(ex)));
    this._launcher.register('voice','Ses','mic',()=>this._open(new VoiceControlApp(ex)));
    this._launcher.register('sd','SD','sd',()=>this._open(new SDHealthApp(ex)));
    this._launcher.register('apt','APT','apt',()=>this._open(new APTUpdateApp(ex)));
    this._launcher.register('todo','Todo','todo',()=>this._open(new TodoApp(ex)));
    if (this._launcher.appCount !== 57) console.error(`Launcher parity error: expected 57 apps, got ${this._launcher.appCount}`);
  }
  _registerTasks(){this._sched.add('statusbar',()=>this._statusbar.update(),1.0);this._sched.add('notif',()=>this._notif.tick(),0.1);this._sched.add('app_update',()=>this._tickApp(),0.05);}
  _bindButtons(){this._input.on(EVT_UP,()=>this._onUp());this._input.on(EVT_DOWN,()=>this._onDown());this._input.on(EVT_SEL,()=>this._onSel());this._input.on(EVT_LONG,()=>this._onLong());}
  _open(app){this._app=app;this._state=State.APP;this._emitState();}
  _goHome(){this._app=null;this._state=State.HOME;this._emitState();}
  _emitState(){if(this._onStateChange){const names=['SPLASH','HOME','APP'];const app=this._app?(this._app.constructor.LABEL||this._app.constructor.NAME||'?'):'Launcher';this._onStateChange(names[this._state],app);}}
  _onUp(){if(this._state===State.HOME)this._launcher.moveUp();else if(this._state===State.APP&&this._app?.onUp)this._app.onUp();}
  _onDown(){if(this._state===State.HOME)this._launcher.moveDown();else if(this._state===State.APP&&this._app?.onDown)this._app.onDown();}
  _onSel(){if(this._state===State.HOME)this._launcher.select();else if(this._state===State.APP&&this._app?.onSel)this._app.onSel();}
  _onLong(){if(this._state===State.APP&&this._app?.onLong)this._app.onLong();else this._goHome();}
  _tickApp(){if(this._app?.update)this._app.update();}
  run(){this._input.start();this._notif.push('57 app simulator hazir!',2.0);this._emitState();const frameMs=1000/TARGET_FPS;let lastFrame=0;const loop=(ts)=>{if(ts-lastFrame>=frameMs){lastFrame=ts;this._sched.tick();this._render();}requestAnimationFrame(loop);};requestAnimationFrame(loop);}
  _render(){const fb=this._fb;fb.clear();if(this._state===State.SPLASH)this._drawSplash();else if(this._state===State.HOME){this._statusbar.draw(fb);this._launcher.draw(fb);}else if(this._state===State.APP&&this._app){this._statusbar.draw(fb);this._app.draw(fb);}if(this._notif.active)this._notif.draw(fb);this._display.render(fb);if(this._state===State.SPLASH){this._splashT++;if(this._splashT>=SPLASH_FRAMES){this._state=State.HOME;this._emitState();}}}
  _drawSplash(){const fb=this._fb,t=this._splashT;if(t<15)fb.circle(64,32,t*2,true,true);else{fb.circle(64,32,30,true,true);fb.textCentered('TinyOLED',20,false);fb.textCentered('57 App Web',30,false);fb.textCentered('Simulator',40,false);}if(t>=20){const progress=Math.min(1,(t-20)/(SPLASH_FRAMES-20)),barW=Math.floor(100*progress);fb.rect(14,56,100,5);if(barW>0)fb.rect(15,57,barW-1,3,true,true);}}
}

File: desktop/shell.py
"""
TinyOLED Desktop — Ana Masaüstü Kabuğu
Tüm bileşenleri bir araya getirir:
  StatusBar + Launcher + App yönetimi + Bildirimler + Zamanlayıcı

Durum makinesi:
  HOME   → Launcher göster
  APP    → Aktif uygulama göster
  SPLASH → Açılış animasyonu
"""

import time
from enum import Enum, auto
from typing import Optional, Any

from core.framebuffer import Framebuffer
from core.display     import SSD1306
from core.input       import ButtonManager, EVT_UP, EVT_DOWN, EVT_SEL, EVT_LONG
from core.scheduler   import Scheduler
from core.font        import Font

from desktop.statusbar    import StatusBar
from desktop.launcher     import Launcher
from desktop.notification import NotificationManager

from apps.clock       import ClockApp
from apps.sysinfo     import SysInfoApp
from apps.wifi        import WiFiApp
from apps.filebrowser import FileBrowserApp
from apps.settings    import SettingsApp
from apps.power       import PowerApp

# ── Yeni Uygulamalar ──────────────────────────────────────────
from apps.snake          import SnakeApp
from apps.flappy         import FlappyApp
from apps.tamagotchi     import TamagotchiApp
from apps.dice           import DiceApp
from apps.graph          import GraphApp
from apps.docker_mon     import DockerMonApp
from apps.systemd_mon    import SystemdMonApp
from apps.gpio_viewer    import GPIOViewerApp
from apps.i2c_scan       import I2CScanApp
from apps.script_runner  import ScriptRunnerApp
from apps.oscilloscope   import OscilloscopeApp
from apps.multimeter     import MultimeterApp
from apps.ssh_alert      import SSHAlertApp
from apps.totp           import TOTPApp
from apps.passgen        import PassGenApp
from apps.wifi_scan      import WiFiScanApp
from apps.ss_starfield   import StarfieldApp
from apps.ss_gameoflife  import GameOfLifeApp
from apps.ss_dvd         import DVDLogoApp
from apps.ss_matrix      import MatrixRainApp
from apps.crypto         import CryptoApp
from apps.cube3d         import Cube3DApp
from apps.fractal        import FractalApp
from apps.qrcode_gen     import QRCodeGenApp
from apps.github_tracker import GitHubTrackerApp
from apps.speedtest_app  import SpeedtestApp
from apps.hackernews     import HackerNewsApp
from apps.sensors        import SensorsApp
from apps.ups_battery    import UPSBatteryApp
from apps.plant_monitor  import PlantMonitorApp
from apps.compass        import CompassApp
from apps.mp3player      import MP3PlayerApp
from apps.radio          import RadioApp
from apps.videoplayer    import VideoPlayerApp
from apps.telegram_bot   import TelegramBotApp
from apps.email_counter  import EmailCounterApp
from apps.moon           import MoonApp
from apps.worldclock     import WorldClockApp
from apps.pixelart       import PixelArtApp
from apps.screenshot     import ScreenshotApp
from apps.pomodoro       import PomodoroApp
from apps.breathing      import BreathingApp
from apps.workout        import WorkoutApp
from apps.servo          import ServoApp
from apps.rc_car         import RCCarApp
from apps.pihole         import PiholeApp
from apps.ipcam          import IPCamApp
from apps.voice          import VoiceApp
from apps.sd_health      import SDHealthApp
from apps.apt_update     import APTUpdateApp
from apps.todo           import TodoApp


class State(Enum):
    SPLASH = auto()
    HOME   = auto()
    APP    = auto()


SPLASH_FRAMES = 40    # ~2 saniye @ 20fps
TARGET_FPS    = 20
FRAME_MS      = 1.0 / TARGET_FPS


class Desktop:
    """
    TinyOLED Desktop ana döngüsü.
    main.py tarafından başlatılır.
    """

    def __init__(self, display: SSD1306):
        self._display  = display
        self._fb       = Framebuffer()
        self._state    = State.SPLASH
        self._app      = None          # Aktif uygulama nesnesi
        self._splash_t = 0

        # Alt sistem bileşenleri
        self._statusbar = StatusBar()
        self._launcher  = Launcher()
        self._notif     = NotificationManager()
        self._sched     = Scheduler()
        self._input     = ButtonManager()

        self._register_apps()
        self._register_tasks()
        self._bind_buttons()

    # ── Uygulama Kaydı ────────────────────────────────────────
    def _register_apps(self):
        d  = self._display
        nx = self._notif.push
        ex = self._go_home

        # ── Orijinal Uygulamalar ──
        self._launcher.register("clock",    "Saat",   "clock",    lambda: self._open(ClockApp(ex)))
        self._launcher.register("sysinfo",  "Sistem", "cpu",      lambda: self._open(SysInfoApp(ex)))
        self._launcher.register("wifi",     "WiFi",   "wifi",     lambda: self._open(WiFiApp(ex, nx)))
        self._launcher.register("files",    "Dosya",  "folder",   lambda: self._open(FileBrowserApp(ex, nx)))
        self._launcher.register("settings", "Ayar",   "gear",     lambda: self._open(SettingsApp(ex, nx, d)))
        self._launcher.register("power",    "Guc",    "power",    lambda: self._open(PowerApp(ex, nx, d)))

        # ── Oyunlar ──
        self._launcher.register("snake",    "Yilan",  "snake",    lambda: self._open(SnakeApp(ex)))
        self._launcher.register("flappy",   "Flappy", "bird",     lambda: self._open(FlappyApp(ex)))
        self._launcher.register("tamagotchi","Pet",   "pet",      lambda: self._open(TamagotchiApp(ex)))
        self._launcher.register("dice",     "Zar",    "dice",     lambda: self._open(DiceApp(ex)))

        # ── İzleme & Grafik ──
        self._launcher.register("graph",    "Graf",   "graph",    lambda: self._open(GraphApp(ex)))
        self._launcher.register("docker",   "Dockr",  "docker",   lambda: self._open(DockerMonApp(ex, nx)))
        self._launcher.register("systemd",  "Servi",  "service",  lambda: self._open(SystemdMonApp(ex, nx)))

        # ── Geliştirici Araçları ──
        self._launcher.register("gpio",     "GPIO",   "pin",      lambda: self._open(GPIOViewerApp(ex)))
        self._launcher.register("i2c",      "I2C",    "i2c",      lambda: self._open(I2CScanApp(ex)))
        self._launcher.register("scripts",  "Komut",  "script",   lambda: self._open(ScriptRunnerApp(ex, nx)))
        self._launcher.register("scope",    "Scope",  "scope",    lambda: self._open(OscilloscopeApp(ex)))
        self._launcher.register("volt",     "Volt",   "volt",     lambda: self._open(MultimeterApp(ex)))

        # ── Güvenlik ──
        self._launcher.register("sshalert", "SSH",    "shield",   lambda: self._open(SSHAlertApp(ex, nx)))
        self._launcher.register("totp",     "2FA",    "key",      lambda: self._open(TOTPApp(ex)))
        self._launcher.register("passgen",  "Sifre",  "passkey",  lambda: self._open(PassGenApp(ex)))
        self._launcher.register("wifiscan", "Tara",   "radar",    lambda: self._open(WiFiScanApp(ex)))

        # ── Ekran Koruyucular ──
        self._launcher.register("starfield","Yldiz",  "star",     lambda: self._open(StarfieldApp(ex)))
        self._launcher.register("life",     "Life",   "cell",     lambda: self._open(GameOfLifeApp(ex)))
        self._launcher.register("dvd",      "DVD",    "dvd",      lambda: self._open(DVDLogoApp(ex)))
        self._launcher.register("matrix",   "Matrx",  "matrix",   lambda: self._open(MatrixRainApp(ex)))

        # ── Finans & 3D ──
        self._launcher.register("crypto",   "Kripto", "crypto",   lambda: self._open(CryptoApp(ex)))
        self._launcher.register("cube3d",   "3D",     "cube",     lambda: self._open(Cube3DApp(ex)))
        self._launcher.register("fractal",  "Fraktl", "fractal",  lambda: self._open(FractalApp(ex)))

        # ── Akıllı Araçlar ──
        self._launcher.register("qrcode",   "QR",     "qr",       lambda: self._open(QRCodeGenApp(ex)))
        self._launcher.register("github",   "Git",    "github",   lambda: self._open(GitHubTrackerApp(ex)))
        self._launcher.register("speedtest","Hiz",    "speed",    lambda: self._open(SpeedtestApp(ex)))
        self._launcher.register("hackernews","HN",    "news",     lambda: self._open(HackerNewsApp(ex)))

        # ── Sensörler ──
        self._launcher.register("sensors",  "Sensr",  "temp",     lambda: self._open(SensorsApp(ex)))
        self._launcher.register("ups",      "UPS",    "ups",      lambda: self._open(UPSBatteryApp(ex)))
        self._launcher.register("plant",    "Bitki",  "plant",    lambda: self._open(PlantMonitorApp(ex)))
        self._launcher.register("compass",  "Pusla",  "compass",  lambda: self._open(CompassApp(ex)))

        # ── Medya ──
        self._launcher.register("mp3",      "Muzik",  "music",    lambda: self._open(MP3PlayerApp(ex)))
        self._launcher.register("radio",    "Radyo",  "radio",    lambda: self._open(RadioApp(ex)))
        self._launcher.register("video",    "Video",  "video",    lambda: self._open(VideoPlayerApp(ex)))

        # ── Mesajlaşma ──
        self._launcher.register("telegram", "Telgm",  "telegram", lambda: self._open(TelegramBotApp(ex)))
        self._launcher.register("email",    "Email",  "email",    lambda: self._open(EmailCounterApp(ex)))

        # ── Astronomi ──
        self._launcher.register("moon",     "Ay",     "moon",     lambda: self._open(MoonApp(ex)))
        self._launcher.register("worldclock","Dunya",  "world",   lambda: self._open(WorldClockApp(ex)))

        # ── Yaratıcı ──
        self._launcher.register("pixelart", "Pixel",  "paint",    lambda: self._open(PixelArtApp(ex)))
        self._launcher.register("screenshot","SS",    "camera",   lambda: self._open(ScreenshotApp(ex, nx)))

        # ── Sağlık & Fitness ──
        self._launcher.register("pomodoro", "Pomo",   "timer",    lambda: self._open(PomodoroApp(ex, nx)))
        self._launcher.register("breath",   "Nefes",  "breath",   lambda: self._open(BreathingApp(ex)))
        self._launcher.register("workout",  "HIIT",   "workout",  lambda: self._open(WorkoutApp(ex, nx)))

        # ── Robotik ──
        self._launcher.register("servo",    "Servo",  "servo",    lambda: self._open(ServoApp(ex)))
        self._launcher.register("rc_car",   "Araba",  "car",      lambda: self._open(RCCarApp(ex)))

        # ── Ağ ──
        self._launcher.register("pihole",   "Pi-h",   "pihole",   lambda: self._open(PiholeApp(ex)))
        self._launcher.register("ipcam",    "Kamera", "camera",   lambda: self._open(IPCamApp(ex)))

        # ── Ses ──
        self._launcher.register("voice",    "Ses",    "mic",      lambda: self._open(VoiceApp(ex, nx)))

        # ── Bakım ──
        self._launcher.register("sdhealth", "SD",     "sd",       lambda: self._open(SDHealthApp(ex)))
        self._launcher.register("apt",      "APT",    "apt",      lambda: self._open(APTUpdateApp(ex, nx)))
        self._launcher.register("todo",     "Gorev",  "todo",     lambda: self._open(TodoApp(ex)))

    def _register_tasks(self):
        self._sched.add("statusbar", self._statusbar.update, interval=1.0)
        self._sched.add("notif",     self._notif.tick,        interval=0.1)
        self._sched.add("app_update",self._tick_app,          interval=1.0)

    def _bind_buttons(self):
        b = self._input
        b.on(EVT_UP,   self._on_up)
        b.on(EVT_DOWN, self._on_down)
        b.on(EVT_SEL,  self._on_sel)
        b.on(EVT_LONG, self._on_long)

    # ── Durum Geçişleri ────────────────────────────────────────
    def _open(self, app):
        self._app   = app
        self._state = State.APP

    def _go_home(self):
        self._app   = None
        self._state = State.HOME

    # ── Buton Olayları ─────────────────────────────────────────
    def _on_up(self):
        if self._state == State.HOME:
            self._launcher.move_up()
        elif self._state == State.APP and hasattr(self._app, "on_up"):
            self._app.on_up()

    def _on_down(self):
        if self._state == State.HOME:
            self._launcher.move_down()
        elif self._state == State.APP and hasattr(self._app, "on_down"):
            self._app.on_down()

    def _on_sel(self):
        if self._state == State.HOME:
            self._launcher.select()
        elif self._state == State.APP and hasattr(self._app, "on_sel"):
            self._app.on_sel()

    def _on_long(self):
        if self._state == State.APP and hasattr(self._app, "on_long"):
            self._app.on_long()
        else:
            self._go_home()

    def _tick_app(self):
        if self._app and hasattr(self._app, "update"):
            self._app.update()

    # ── Ana Döngü ──────────────────────────────────────────────
    def run(self):
        """Bloklayan ana döngü. Ctrl+C ile durdurulur."""
        self._input.start()
        self._notif.push("TinyOLED v1.0 hazir!", duration=2.0)

        try:
            while True:
                frame_start = time.monotonic()

                self._sched.tick()
                self._render()

                elapsed = time.monotonic() - frame_start
                sleep_t = max(0.0, FRAME_MS - elapsed)
                if sleep_t > 0:
                    time.sleep(sleep_t)

        except KeyboardInterrupt:
            pass
        finally:
            self._fb.clear()
            self._fb.flush(self._display)
            self._display.power(False)
            self._input.stop()

    # ── Render ─────────────────────────────────────────────────
    def _render(self):
        fb = self._fb
        fb.clear()

        if self._state == State.SPLASH:
            self._draw_splash()
        elif self._state == State.HOME:
            self._statusbar.draw(fb)
            self._launcher.draw(fb)
        elif self._state == State.APP and self._app:
            self._statusbar.draw(fb)
            self._app.draw(fb)

        if self._notif.active:
            self._notif.draw(fb)

        fb.flush(self._display)

        # Splash'ten HOME'a geçiş
        if self._state == State.SPLASH:
            self._splash_t += 1
            if self._splash_t >= SPLASH_FRAMES:
                self._state = State.HOME

    def _draw_splash(self):
        fb = self._fb
        t  = self._splash_t

        # Logo animasyonu: dairesel açılma
        if t < 15:
            r = t * 2
            fb.circle(64, 32, r, fill=True)
        else:
            fb.circle(64, 32, 30, fill=True)
            # Logo metni (siyah üzerine beyaz)
            fb.text_centered("TinyOLED", 22, on=False)
            fb.text_centered("Desktop v1.0", 31, on=False)
            fb.text_centered("Pi 3B+", 40, on=False)

        # Alt yükleme çubuğu
        if t >= 20:
            progress = min(1.0, (t - 20) / (SPLASH_FRAMES - 20))
            bar_w    = int(100 * progress)
            fb.rect(14, 56, 100, 5)
            if bar_w > 0:
                fb.rect(15, 57, bar_w - 1, 3, fill=True)
